### DISCUSSION TITLE
Problem with Authorization

### DISCUSSION THREAD

**Discussion by eweitlaner** [bimmerconnected/bimmer_connected#597] [Q&A]
I try to connect to BMW Connected Drive with your code, the problem starts with the second call to the authenticate Endpoint, I never get a response code 302, but always code 401. It looks like the authorization from the first step is invalid ?!

The authorization from the first step looks like

WaaZupo1x0DR27T61AGth_yleXI.\*AAJTSQACMFFFAlNLABw4ZkZFK3Qza3pqV3lIVTVMdk1odThuRHNSWEk9AAR0eXBlAANDVFMAAlMxAAssMQ..\*

(I changed some letters), is the ending '..*' correct? Any idea how to find the problem?

Thank you.
BR erwin

**Comment by rikroe** [comment]
Are you trying to rebuild the login steps by yourself?

However the ending with `..*` seems to be correct, I see the same. Are you sure you are using the correct URL/headers/form data?

**Reply by eweitlaner** [reply]
Ok, the authorization seems right, thank you Richard. So it has to be URL/headers or form data? The 'harder' step to get the authorization with username and password works, for the second step I just remove user, password, and grant_type from form data and add the authorization from step 1. Any mistake?

Thank you, BR erwin

**Reply by rikroe** [reply]
It has to be a combination of URL, headers and form data.

And yes, for the second step you can remove user/password/grant_type from form data. Make sure that you are using the right endpoint (`/token`) and that you add the `client-version` header (I think this is required as well).

**Reply by eweitlaner** [reply]
Are you sure to change the url for the second call?

**Reply by rikroe** [reply]
Sorry, mixed it up. Both first and second request need to go to `/authenticate`, only third request to `/token`

**Reply by eweitlaner** [reply]
3 more hours spent, unfortunately no chance, I tried to check every part, I managed to get the following error:

HttpPost Status: 401
{"error": "invalid_client", "error_description": "Client authentication failed (e.g., login failure, unknown client, no client authentication included or unsupported authentication method)"}

Sounds like wrong client. I am not sure if it could be the client-version. Could you please check the value 'android(TQ2A.230405.003.B2);bmw;3.9.0(27760);row' for Rest of the World (Austria ,-)) .. should the client-version be delivered as header, url parameter or form data?

Thank you in advance.

**Reply by rikroe** [reply]
This is the default error for everything related to authentication. Is also appears if you are rate-limited on authorization (can happen, just wait 24 hours) or if your password is wrong (e.g. because you used `´` which at some point got translated to `'` by the Javascript before sending it to the server).

Alternatively, check the login flow of [TA2k/ioBroker.bmw](https://github.com/TA2k/ioBroker.bmw/blob/master/main.js#L214) which does it slightly differently.

Client ID looks good to me.
